### PR TITLE
[docs][tools] add GenerateDocsAPIData command, add APISection docs component

### PR DESCRIPTION
**Pull request by Simek**
# Why & How

Keeping API/components docs up to date might be a painful and challenging process. The goal of `@expo/docs-generator` is to speed up, improve the process and help developers to focus more on code than documentation.

This is a first PR of upcoming series of PRs which includes:
* new Expo Tools command `GenerateDocsAPIData` for generating the JSON data for components
  * the types extraction is based on the [TypeDoc](https://typedoc.org/) tool and uses JSON output from the TS types
* basic version of `APISection` component
* very basic versioning support

Currently, the generated API section is used only for `mail-composer` and outputs all the same content which was manually typed in and additionally includes few more types definitions and one enum.

The UI and content arrangement isn't perfect yet, there are small issues like, for example `InlineCode` headers are not displaying correctly on the right sidebar. Also I have put some work into content arrangement, it was mainly based on existing pages, but not all of them follows the same pattern, so probably this part also will change in the future.

Not all sub-sections and types of API data are currently supported, but the current set was enough to cover `mail-composer` API.

I'm also experiencing issue in dev mode, which result in duplicated right sidebar links, when page was auto refresh due to file content changes.

Fixes ENG-742

# Test Plan

I have tested the changes on `localhost` but they might be more changes required to make this production worthy.

# Preview

docs-gen-preview

CC @brentvatne @bbarthec

### CODE REVIEW

File: docs/components/plugins/APISection.tsx [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+              <td>{typeProperty.comment ? typeProperty.comment.shortText : '-'}</td>
+            </tr>
+          ))}
+        </tbody>
+      </table>
+    </>
+  );
+};
+
+const processData = (packageName: string, version: string = 'unversioned'): JSX.Element => {
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx:161]:
```suggestion
const renderAPI = (packageName: string, version: string = 'unversioned'): JSX.Element => {
```
I would rather go with naming function after what's happening in it 😉

File: docs/components/plugins/APISection.tsx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  const data = require(`~/public/static/data/${version}/${packageName}.json`);
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx]: Let's inject data into this method to make it `pure` render method.

File: docs/components/plugins/APISection.tsx [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  const methods = data.children.filter(g => g.kind === 64);
+  const types = data.children.filter(g => g.kind === 4194304);
+  const enums = data.children.filter(g => g.kind === 4);
+
+  // TODO: Props, Constants, Static Methods and probably few more sections
+
+  return (
+    <>
+      {methods ? <H2>Methods</H2> : null}
+      {methods ? methods.map(renderMethod) : null}
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx]:
```suggestion
      <Methods data={methodsData} />
```

File: docs/components/plugins/APISection.tsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      {types ? <H2>Types</H2> : null}
+      {types ? types.map(renderType) : null}
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx]:
```suggestion
      <Types data={typesData} />
```

File: docs/components/plugins/APISection.tsx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+      {enums ? <H2>Enums</H2> : null}
+      {enums ? enums.map(renderEnum) : null}
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx]:
```suggestion
      <Enums data={enumsData} />
```

File: docs/components/plugins/APISection.tsx [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+  );
+};
+
+const processData = (packageName: string, version: string = 'unversioned'): JSX.Element => {
+  try {
+    const data = require(`~/public/static/data/${version}/${packageName}.json`);
+
+    const methods = data.children.filter(g => g.kind === 64);
+    const types = data.children.filter(g => g.kind === 4194304);
+    const enums = data.children.filter(g => g.kind === 4);
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx:167]: We definitely need to convert this numeric values into constants. Maybe these are available via some ts-doc-related package? 🤔 If no, we need to define them ourselves.
Best approach would be to have a method:
```
function getData(data: JSONData, dataType: DataTypeEnum): SpecificDataTypeSet
```

File: docs/components/plugins/APISection.tsx [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    // TODO: Props, Constants, Static Methods and probably few more sections
+
+    return (
+      <>
+        {methods ? <H2>Methods</H2> : null}
+        {methods ? methods.map(renderMethod) : null}
+        {types ? <H2>Types</H2> : null}
+        {types ? types.map(renderType) : null}
+        {enums ? <H2>Enums</H2> : null}
+        {enums ? enums.map(renderEnum) : null}
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx:178]:
```suggestion
        <Methods  data={methodsData} />
        <Types data={typesData} />
        <Enums data={enumData} />
```

File: docs/components/plugins/APISection.tsx [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  } else if (elementType.name) {
+    if (type === 'array') {
+      return elementType.name + '[]';
+    }
+    return elementType.name + type;
+  }
+  return 'undefined';
+};
+
+const renderMethod = ({ signatures }: any): JSX.Element =>
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx:58]: Let's convert it into Component, shall we? 🤔 and extract to separate file 😉

File: docs/components/plugins/APISection.tsx [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        ) : null}
+        {comment?.returns ? (
+          <ReactMarkdown renderers={renderers}>{comment.returns}</ReactMarkdown>
+        ) : null}
+        <hr />
+      </>
+    );
+  });
+
+const renderEnum = ({ name, children, comment }: any): JSX.Element => {
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx:98]: Let's convert it into Component, shall we? 🤔

File: docs/components/plugins/APISection.tsx [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+            </InlineCode>
+            {comment ? ` - ${comment.shortText}` : null}
+          </LI>
+        ))}
+      </UL>
+    </>
+  );
+};
+
+const renderType = ({ name, comment, type }: any): JSX.Element => {
```
Comment (bbarthec) [inline docs/components/plugins/APISection.tsx:118]: Let's convert it into Component, shall we? 🤔

**Review by bbarthec — CHANGES_REQUESTED** [review]
Looks promising, but I want you to perform some changes:
- Let's move this new package with single command into `tools` directory
- Let's remove `react-markdown` and handle these cases manually. I'm not sure why you've introduced this dependency here, because you're using it only for blocks of text. If there are scenario when we need to handle `list` or `link` we can manage on our own I believe? 🤔 I'm open to discussion anyway 😉

File: yarn.lock [hunk 11]
```diff
@@ -16713,6 +16760,39 @@ react-native-webview@11.2.3:
     escape-string-regexp "2.0.0"
     invariant "2.2.4"
 
+react-native@0.63.2:
```
[the remark this answers is not in the document]
Comment (Simek) [reply yarn.lock]: I was having issues with submodules @ide helped me solve them, it was the remnant in code which was removed in `5ae04b1`.

**Comment by Simek** [comment]
> * Let's remove `react-markdown` and handle these cases manually. I'm not sure why you've introduced this dependency here, because you're using it only for blocks of text. If there are scenario when we need to handle `list` or `link` we can manage on our own I believe? 🤔 I'm open to discussion anyway 😉

`react-markdown` is required, otherwise content/comment output is invalid + it enables to apply the Expo themed components easily:
Screenshot 2021-03-18 at 17 26 52

File: docs/components/plugins/APISection.tsx (hunk 6 above)
Comment (Simek) [reply docs/components/plugins/APISection.tsx]: There are types build in `typedoc` (`ReflectionKind`), but to utilize them we will need to import whole TypeDoc package as `dependency` to the `docs`which is a bit of overkill:
* https://github.com/TypeStrong/typedoc/blob/master/src/lib/models/reflections/abstract.ts#L37-L64

Other TypeDoc users seems also to skip this and creating there own types in-place:
* https://github.com/jquense/docpocalypse/blob/master/packages/theme/src/components/typedoc-types.ts#L4-L38

Comment (Simek) [reply docs/components/plugins/APISection.tsx]: Also the included in TypeDoc `JSONOutput` schema is very fragile and unpleasant to use, so for now (PoC) I have opted out dealing with this problem:
* https://github.com/TypeStrong/typedoc/issues/789
* https://github.com/TypeStrong/typedoc/issues/605

File: packages/expo-mail-composer/src/MailComposer.ts [hunk 12]
```diff
@@ -1,15 +1,20 @@
 import ExpoMailComposer from './ExpoMailComposer';
 import { MailComposerOptions, MailComposerResult } from './MailComposer.types';
 
+/**
+ * Opens a mail modal for iOS and a mail app intent for Android and fills the fields with provided data. On iOS you will need to be signed into the Mail app.
+ * @return Resolves to a promise with object containing status field that could be either sent, saved or cancelled. Android does not provide such info so it always resolves to sent.
+ */
```
[the remark this answers is not in the document]
Comment (ide) [reply packages/expo-mail-composer/src/MailComposer.ts:7]: Wrap to 100 cols, editors have extensions for this

File: packages/expo-mail-composer/src/MailComposer.ts [hunk 13]
```diff
@@ -1,15 +1,20 @@
 import ExpoMailComposer from './ExpoMailComposer';
 import { MailComposerOptions, MailComposerResult } from './MailComposer.types';
 
+/**
+ * Opens a mail modal for iOS and a mail app intent for Android and fills the fields with provided data. On iOS you will need to be signed into the Mail app.
+ * @return Resolves to a promise with object containing status field that could be either sent, saved or cancelled. Android does not provide such info so it always resolves to sent.
```
Comment (ide) [inline packages/expo-mail-composer/src/MailComposer.ts:6]: Now that this documentation is extra visible, technical correctness and quality of writing matter even more:

```suggestion
 * @return A promise fulfilled with an object containing a `status` field that specifies whether an email was sent, saved, or cancelled. Android does not provide this info, so the status is always set as if the email were sent.
```

(@brentvatne I think we will need higher expectations for authors and reviewers for our docblocks now. High-quality docs reflect upon our team's competence.)

**Review by ide — COMMENTED** [review]
I didn't do a thorough review or leave comments everywhere but several sections of prose need to be revised before publishing.

**Comment by Simek** [comment]
@bbarthec You might want to check the code again and re-review this PR 🙂 

FYI: Also, as a workaround (?), I had to add `typedoc` to the workspace `package.json`. When `typedoc` was listed in `tools` dependencies commands start to failing (aster the `docs-generator` package removal). Not sure if it's correct, but it was the last thing which prevents me from pushing the refactor. Please let me know if you have other ideas or fix for this problem.

File: docs/components/plugins/APISection.tsx (hunk 2 above)
Comment (Simek) [reply docs/components/plugins/APISection.tsx]: @bbarthec I think it's bit too early for such optimisations, the code is changing and evolving rapidly, with every new package tested or extracted, but I'm happy to work on refactor when we will be closer to the merge 😉

**Comment by Simek** [comment]
My previous comment about `typedoc` dependency location is no longer correct. I have refactored the generation Command to use `typedoc` module API instead of executing it as CLI via child processes, thanks to @brentvatne for the hint! 🙂

**Review by bbarthec — CHANGES_REQUESTED** [review]
Bunch of new ideas:
- when this PR will be able to land, let's extract `expo-mail-composer` to the separate PR ok?
- get rid of any `any` - all of them come from parts of this imported JSON file and we're able to resolve it quite simply by writing type for this imported data.

**Comment by Simek** [comment]
> * when this PR will be able to land, let's extract `expo-mail-composer` to the separate PR ok?

It's already extracted - #12278 - I'm waiting for merge to cleanup this PR 🙂

There are two more PRs for other packages, I have referenced them here and listed in Statuses gist.

> * get rid of any `any` - all of them come from parts of this imported JSON file and we're able to resolve it quite simply by writing type for this imported data.

I left the JSON types for the end, because the types are changing almost with each commit, but I can start defining them 😉

**Comment by Simek** [comment]
The structure of generate data command is a leftover of refactor from CLI to module API, I will rewrite it as you suggested.

File: tools/src/commands/GenerateDocsAPIData.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+      }
+    );
+  }
+}
+
+export default (program: Command) => {
+  program
+    .command('generate-docs-api-data')
+    .alias('gdad')
+    .description(`Extract API data for docs using TypeDoc.`)
```
[the remark this answers is not in the document]
Comment (Simek) [reply tools/src/commands/GenerateDocsAPIData.ts:89]: Paths are visible in the command output, so I think it's not necessary to duplicate this information. It's already a crowded a bit log 😉 

Screenshot 2021-03-24 at 15 25 45

[the remark this answers is not in the document]
Comment (bbarthec) [reply tools/src/commands/GenerateDocsAPIData.ts:86]: You can always limit down the amount of output going by the rule: if something works, keep the amount of logs to minimum (for example, use [ora spinner](https://www.npmjs.com/package/ora): [here example](https://github.com/expo/expo/pull/11919/files?file-filters%5B%5D=.ts#diff-f6a2b5b52f1ef1302a76f495fca32faf0c32c0e5fc671cf4c56fb394ba03bb92R81-R92)) and when something fails, then provide as much info as possible 😉 .
I guess you don't have to optimise at this point, but you can keep that in mind.

**Review by bbarthec — COMMENTED** [review]
I really like the way it's going 👍 
I wish we could plug-in & play with it on master right away 😉

What would you say about creating some trigger for enabling this generation mechanism by passing some `ENABLE_NEW_DOCS_GENERATION=1` env variable and moving the code to the `**/experimental/**` directory?
That would provide us with an easy way for testing and proceeding with the work more incrementally on master branch with small PRs & adjustments (at this point this PR is growing huge 😅).

If we had this switch working, I would happily merge the generation mechanism right now 😉 
And once we're ready to completely switch to this new way of docs generation we can just remove this `trigger/switch/env variable` and move all the code to the respective directories.

It's up to you, but I would gladly involve some other folks to work on the `@needsAudit` flag or filling the missing doc parts from the next month.

Additionally, I've detected that when we have multiple headings with the very same text (`Returns`/`Arguments` sub-headings in almost every method) then URL for such section ends with `returns-1` or similar. I think that maybe we can handle it differently? Maybe there's option to add parent section reference in such URL? 🤔 cc @brentvatne

**Comment by Simek** [comment]
> What would you say about creating some trigger for enabling this generation mechanism by passing some `ENABLE_NEW_DOCS_GENERATION=1` env variable and moving the code to the `**/experimental/**` directory?
> That would provide us with an easy way for testing and proceeding with the work more incrementally on master branch with small PRs & adjustments (at this point this PR is growing huge 😅).

I think there is no good way to implement a switch, but the generation can be controlled from the command file. Unfortunately making autogenerated docs available it's not one step process, the MD files need to be edited manually, and part of the content need to be deleted, which is an another problem for the switch implementation.

> Additionally, I've detected that when we have multiple headings with the very same text (`Returns`/`Arguments` sub-headings in almost every method) then URL for such section ends with `returns-1` or similar. I think that maybe we can handle it differently? Maybe there's option to add parent section reference in such URL? 🤔 cc @brentvatne

I'm not the author of the code responsible for that, slugs are autogenerated, and the problem it's already occurring in the docs, for example:
* https://docs.expo.io/versions/v40.0.0/sdk/application/#returns-1

I can try to improve the slug generating code later.

**Review by bbarthec — APPROVED** [review]
As for me, it looks very good 👍 
I believe nothing is blocking us from shipping this right now 👍

**Comment by brentvatne** [comment]
this is looking exceptional! on the pages where it's used, it represents a massive improvement over the current state.

the only thing i would like to see change from my first pass over this is to add argument names to the method header:

### currently it's like this

Screen Shot 2021-04-01 at 5 21 41 PM

### i'd like to see it like this

Screen Shot 2021-04-01 at 5 22 32 PM

<hr />

great work on this @Simek! 🎉

**Review by brentvatne — CHANGES_REQUESTED** [review]
let's update the methods to include argument names as mentioned above, and also backport these changes to sdk 41 docs so we can try this live on the beta release

**Comment by Simek** [comment]
@bradvatne I have added the params to the headings and verified that the generated slugs are the same.
Screenshot 2021-04-02 at 13 02 06

Also, as requested, the converted pages have been backported to SDK 41 docs. 🙂

**Review by byCedric — APPROVED** [review]
Overall looks awesome!

I do have some concerns related to performance. We have some `require` and data manipulations in the react components themselves. But if they cause a lot of stress on the performance, we can always improve that in follow up PRs 😄

**Review by brentvatne — APPROVED** [review]

**Pull request merged**